File: src/com/avail/optimizer/values/L2SemanticSlot.java
/**
 * L2SemanticSlot.java
 * Copyright © 1993-2017, The Avail Foundation, LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package com.avail.optimizer.values;
import com.avail.descriptor.A_Continuation;
import com.avail.utility.evaluation.Transformer1NotNull;

import static com.avail.descriptor.AvailObject.multiplier;

/**
 * A semantic value which represents a slot of some {@link Frame}'s effective
 * {@link A_Continuation}.
 */
final class L2SemanticSlot extends L2SemanticValue
{
	/** The {@link Frame} for which this is a virtualized slot. */
	public final Frame frame;

	/** The one-based index of the slot in its {@link Frame}. */
	public final int slotIndex;

	/**
	 * Create a new {@code L2SemanticSlot} semantic value.
	 *
	 * @param frame
	 *        The frame for which this represents a slot of a virtualized
	 *        continuation.
	 * @param slotIndex
	 *        The one-based index of the slot in that frame.
	 */
	L2SemanticSlot (final Frame frame, final int slotIndex)
	{
		this.frame = frame;
		this.slotIndex = slotIndex;
	}

	@Override
	public boolean equals (final Object obj)
	{
		if (!(obj instanceof L2SemanticSlot))
		{
			return false;
		}
		final L2SemanticSlot slot = (L2SemanticSlot) obj;
		return frame.equals(slot.frame)
			&& slotIndex == slot.slotIndex;
	}

	@Override
	public int hashCode ()
	{
		return (frame.hashCode() ^ slotIndex) * multiplier;
	}

	@Override
	public L2SemanticSlot transform (
		final Transformer1NotNull<L2SemanticValue, L2SemanticValue>
			semanticValueTransformer,
		final Transformer1NotNull<Frame, Frame> frameTransformer)
	{
		final Frame newFrame = frameTransformer.value(frame);
		return newFrame.equals(frame)
			? this
			: new L2SemanticSlot(newFrame, slotIndex);
	}

	@Override
	public String toString ()
	{
		return "Slot #" + slotIndex + " of " + frame;
	}
}

File: avail/src/main/kotlin/avail/optimizer/values/L2SemanticSlot.kt
/*
 * L2SemanticSlot.kt
 * Copyright © 1993-2022, The Avail Foundation, LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package avail.optimizer.values

import avail.descriptor.functions.A_Continuation
import avail.descriptor.representation.AvailObject
import avail.interpreter.levelTwo.register.BOXED_KIND

/**
 * A semantic value which represents a slot of some [Frame]'s effective
 * [A_Continuation].
 *
 * @author Mark van Gulik &lt;mark@availlang.org&gt;
 * @property slotIndex
 *   The one-based index of the slot in its [Frame].
 * @property pcAfter
 *   The level one [A_Continuation.pc] at the position just after the nybblecode
 *   instruction that produced this value.  This serves to distinguish semantic
 *   slots at the same index but at different times, allowing a correct SSA
 *   graph and the reordering that it supports.
 *
 * @constructor
 * Create a new `L2SemanticSlot` semantic value.
 *
 * @param frame
 *   The frame for which this represents a slot of a virtualized continuation.
 * @param slotIndex
 *   The one-based index of the slot in that frame.
 * @param pcAfter
 *   The L1 program counter just after the instruction responsible for
 *   effectively writing this value in the frame's slot.
 * @param optionalName
 *   Either a [String] that provides a useful naming hint for this slot, or
 *   `null`.
 */
internal class L2SemanticSlot constructor(
	frame: Frame,
	val slotIndex: Int,
	val pcAfter: Int,
	private val optionalName: String?
) : L2FrameSpecificSemanticValue(
	frame, slotIndex * AvailObject.multiplier xor pcAfter)
{
	init
	{
		assert(slotIndex >= 1)
	}

	// Note: Ignore the optionalName.
	override fun equalsSemanticValue(other: L2SemanticValue<*>) =
		(other is L2SemanticSlot
			&& super.equalsSemanticValue(other)
			&& slotIndex == other.slotIndex
			&& pcAfter == other.pcAfter)

	override fun transform(
		semanticValueTransformer:
			(L2SemanticValue<BOXED_KIND>) -> L2SemanticValue<BOXED_KIND>,
		frameTransformer: (Frame) -> Frame
	): L2SemanticBoxedValue =
		frameTransformer(frame()).let {
			if (it == frame()) this
			else L2SemanticSlot(it, slotIndex, pcAfter, optionalName)
		}

	override fun primaryVisualSortKey() = PrimaryVisualSortKey.SLOT

	override fun toString(): String = buildString {
		when (optionalName)
		{
			null -> append(slotIndex)
			else -> append(optionalName)
		}
		append("-")
		append(pcAfter)
		if (frame.depth() > 1) append("[$frame]")
	}


	override fun toStringForSynonym(): String = buildString {
		when (optionalName)
		{
			null -> append(slotIndex)
			else -> append(optionalName)
		}
		append("-")
		append(pcAfter)
		if (frame.depth() > 1) append("[$frame]")
	}
}
